homeassistant/components/sensor/geo_rss_events.py

```diff
@@ -141,22 +207,70 @@ def update(self):
             _LOGGER.debug("Adding events to sensor %s: %s", self.entity_id,
                           my_events)
             self._state = len(my_events)
+            # Sort events if configured to do so.
+            if self._sort_by is not None:
+                min_object = MinType()
+                my_events = sorted(my_events,
+                                   key=lambda event:
+                                   min_object if self._sort_by not in event or
+                                   event[self._sort_by] is None
+                                   else event[self._sort_by],
+                                   reverse=self._sort_reverse)
             # And now compute the attributes from the filtered events.
             matrix = {}
             for event in my_events:
                 matrix[event[ATTR_TITLE]] = '{:.0f}km'.format(
                     event[ATTR_DISTANCE])
             self._state_attributes = matrix
+            # Finally publish new events to the bus.
+            if self._publish_events:
+                events_to_publish = self.filter_events_to_publish(my_events)
+                _LOGGER.debug("New events to publish: %s", events_to_publish)
+                self.publish_events(events_to_publish)
```

Comment:
There should never be side effects from calling update. There is no reason to fire events because the state is written to the state machine.

Reply:
If this service is based on events to begin with, it shouldn't have been written as an entity.

Reply:
Thanks for the review, @balloob.
Sorry, I wasn't fully aware of the implications of firing events from an entity. I was loosely following the `crimereports` sensor's implementation which is similar to this sensor, and is also written as an entity and fire events to the bus.
Anyway, I think if I start removing the event firing code, there may not be too much left at the end.
My original intention was to have the ability to retrieve data from the external GeoRSS feed, process the data in HA, show some information in the UI, and also provide more detailed information to the event bus, so that for example an automation could alert users about these external events.
So, would a better approach be to refactor the code now, and introduce a new component that takes care of retrieving the GeoRSS feed. This component could then fire events and feed the sensor, while the sensor code itself becomes very lean is really only responsible for displaying the data in the UI.

Reply:
Yes, it would be a better approach to introduce a new component now that will take care of firing the events.